File: packages/sdkwork-utils-java/src/main/java/com/sdkwork/utils/CurrencyUtils.java
package com.sdkwork.utils;

import java.util.Locale;
import java.util.Map;

public final class CurrencyUtils {
    private static final Map<String, CurrencyMeta> KNOWN = Map.ofEntries(
            Map.entry("USD", new CurrencyMeta(2, "$")),
            Map.entry("EUR", new CurrencyMeta(2, "€")),
            Map.entry("GBP", new CurrencyMeta(2, "£")),
            Map.entry("CNY", new CurrencyMeta(2, "¥")),
            Map.entry("JPY", new CurrencyMeta(0, "¥")),
            Map.entry("KRW", new CurrencyMeta(0, "₩")),
            Map.entry("HKD", new CurrencyMeta(2, "HK$")),
            Map.entry("TWD", new CurrencyMeta(2, "NT$")),
            Map.entry("CHF", new CurrencyMeta(2, "CHF")),
            Map.entry("CAD", new CurrencyMeta(2, "CA$")),
            Map.entry("AUD", new CurrencyMeta(2, "A$")),
            Map.entry("INR", new CurrencyMeta(2, "₹")),
            Map.entry("BHD", new CurrencyMeta(3, "BHD")),
            Map.entry("KWD", new CurrencyMeta(3, "KWD"))
    );

    private CurrencyUtils() {
    }

    private record CurrencyMeta(int exponent, String symbol) {
    }

    private static CurrencyMeta lookup(String code) {
        if (code == null) {
            return null;
        }
        String normalized = code.trim();
        if (normalized.length() != 3 || !normalized.equals(normalized.toUpperCase(Locale.ROOT))
                || !normalized.chars().allMatch(Character::isLetter)) {
            return null;
        }
        return KNOWN.get(normalized);
    }

    public static boolean isCurrencyCode(String value) {
        return lookup(value) != null;
    }

    public static Integer minorUnitExponent(String code) {
        CurrencyMeta meta = lookup(code);
        return meta == null ? null : meta.exponent();
    }

    public static Long toMinorUnits(double amount, String code) {
        CurrencyMeta meta = lookup(code);
        if (meta == null) {
            return null;
        }
        double factor = Math.pow(10, meta.exponent());
        return (long) NumberUtils.round(amount * factor, 0);
    }

    public static Double fromMinorUnits(long minor, String code) {
        CurrencyMeta meta = lookup(code);
        if (meta == null) {
            return null;
        }
        double factor = Math.pow(10, meta.exponent());
        return minor / factor;
    }

    private static boolean suffixLocale(String locale) {
        String normalized = locale.toLowerCase(Locale.ROOT);
        return normalized.startsWith("de")
                || normalized.startsWith("fr")
                || normalized.startsWith("it")
                || normalized.startsWith("es");
    }

    public static String formatCurrency(double amount, String code, String locale) {
        CurrencyMeta meta = lookup(code);
        if (meta == null) {
            return null;
        }
        String formatted = I18nUtils.formatNumberLocale(amount, locale, meta.exponent());
        return suffixLocale(locale) ? formatted + " " + meta.symbol() : meta.symbol() + formatted;
    }
}

File: packages/sdkwork-utils-php/src/CurrencyUtils.php
<?php

declare(strict_types=1);

namespace Sdkwork\Utils;

final class CurrencyUtils
{
    /** @var array<string, array{int, string}> */
    private const KNOWN = [
        'USD' => [2, '$'],
        'EUR' => [2, '€'],
        'GBP' => [2, '£'],
        'CNY' => [2, '¥'],
        'JPY' => [0, '¥'],
        'KRW' => [0, '₩'],
        'HKD' => [2, 'HK$'],
        'TWD' => [2, 'NT$'],
        'CHF' => [2, 'CHF'],
        'CAD' => [2, 'CA$'],
        'AUD' => [2, 'A$'],
        'INR' => [2, '₹'],
        'BHD' => [3, 'BHD'],
        'KWD' => [3, 'KWD'],
    ];

    public static function isCurrencyCode(string $value): bool
    {
        return self::lookup($value) !== null;
    }

    public static function minorUnitExponent(string $code): ?int
    {
        $meta = self::lookup($code);
        return $meta[0] ?? null;
    }

    public static function toMinorUnits(float $amount, string $code): ?int
    {
        $meta = self::lookup($code);
        if ($meta === null) {
            return null;
        }
        [$exponent] = $meta;
        $factor = 10 ** $exponent;
        return (int) NumberUtils::round($amount * $factor, 0);
    }

    public static function fromMinorUnits(int $minor, string $code): ?float
    {
        $meta = self::lookup($code);
        if ($meta === null) {
            return null;
        }
        [$exponent] = $meta;
        $factor = 10 ** $exponent;
        return $minor / $factor;
    }

    public static function formatCurrency(float $amount, string $code, string $locale): ?string
    {
        $meta = self::lookup($code);
        if ($meta === null) {
            return null;
        }
        [$exponent, $symbol] = $meta;
        $formatted = I18nUtils::formatNumberLocale($amount, $locale, $exponent);
        if (self::suffixLocale($locale)) {
            return $formatted . ' ' . $symbol;
        }
        return $symbol . $formatted;
    }

    private static function lookup(string $code): ?array
    {
        $normalized = trim($code);
        if (strlen($normalized) !== 3 || $normalized !== strtoupper($normalized) || !ctype_alpha($normalized)) {
            return null;
        }
        return self::KNOWN[$normalized] ?? null;
    }

    private static function suffixLocale(string $locale): bool
    {
        $normalized = strtolower($locale);
        return str_starts_with($normalized, 'de')
            || str_starts_with($normalized, 'fr')
            || str_starts_with($normalized, 'it')
            || str_starts_with($normalized, 'es');
    }
}
